src/frontend/src/scheduler/plan_fragmenter/pre_fragmenter.rs

```diff
@@ -0,0 +1,610 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_recursion::async_recursion;
+use itertools::Itertools;
+use risingwave_common::buffer::{Bitmap, BitmapBuilder};
+use risingwave_common::catalog::TableDesc;
+use risingwave_common::error::RwError;
+use risingwave_common::hash::{ParallelUnitId, ParallelUnitMapping, VirtualNode};
+use risingwave_common::util::scan_range::ScanRange;
+use risingwave_connector::source::{ConnectorProperties, SplitEnumeratorImpl, SplitImpl};
+use risingwave_pb::batch_plan::ExchangeInfo;
+
+use super::{
+    ExecutionPlanNode, Query, QueryId, QueryStage, QueryStageRef, SchedulerError, SourceScanInfo,
+    StageGraph, StageId, TablePartitionInfo, TableScanInfo,
+};
+use crate::catalog::catalog_service::CatalogReader;
+use crate::optimizer::plan_node::PlanNodeType;
+use crate::optimizer::property::Distribution;
+use crate::optimizer::PlanRef;
+use crate::scheduler::worker_node_manager::WorkerNodeManagerRef;
+use crate::scheduler::SchedulerResult;
+
+/// `BatchPlanFragmenter` splits a query plan into fragments and generate the [`PreStageGraph`].
+pub struct BatchPlanFragmenter {
+    query_id: QueryId,
+    pre_stage_graph_builder: PreStageGraphBuilder,
+    next_stage_id: StageId,
+    worker_node_manager: WorkerNodeManagerRef,
+    catalog_reader: CatalogReader,
+}
+
+impl BatchPlanFragmenter {
+    pub fn new(worker_node_manager: WorkerNodeManagerRef, catalog_reader: CatalogReader) -> Self {
+        Self {
+            query_id: Default::default(),
+            pre_stage_graph_builder: PreStageGraphBuilder::new(),
+            next_stage_id: 0,
+            worker_node_manager,
+            catalog_reader,
+        }
+    }
+}
+
+struct SourceConnectorInfo {
+    pub connector_properties: ConnectorProperties,
+    pub time_bound: (Option<i64>, Option<i64>),
+}
+
+pub struct PreQueryStage {
```

Comment:
Why we need a `PreQueryStage`? Leaving the table scan/source split info as `None` in `QueryStage` isn't enough?